controller/instance_manager_controller.go

```diff
@@ -411,6 +415,22 @@ func (imc *InstanceManagerController) syncStatusWithPod(im *longhorn.InstanceMan
 	return nil
 }
 
+// syncInstanceStatus sets the status of instances in special cases independent of InstanceManagerMonitor (e.g. when
+// InstanceManagerMonitor isn't running yet).
+func (imc *InstanceManagerController) syncInstanceStatus(im *longhorn.InstanceManager) error {
+	if im.Status.CurrentState == longhorn.InstanceManagerStateStarting {
+		// In this state, we can be sure instance processes are not running.
+		// This step prevents other controllers from being confused by stale information.
+		// InstanceManagerMonitor will change this when it polls for the first time.
+		for k, v := range im.Status.Instances {
+			// Retain historical information (e.g. ErrorMsg).
```

Comment:
We might miss the `InstanceManagerStateStarting` when we miss `PodPending` event of the instance-manager pod. In this case, we will not adjust the instances' state correctly. This should be super rare, though. 

I am wondering if we should take action when the instance-manager resource enters `InstanceManagerStateError` state instead of `InstanceManagerStateStarting` state. In this case, we can clear out the `im.Status.Instances` map because that map is no longer valid to use. 

What do you think?

Reply:
I thought about the "missing PodPending" scenario and came to a similar conclusion. However I wasn't particular confident in my understanding of how `InstanceManagerStateError` was being used, so I decided to go for a fix that was as unobtrusive as possible. Since you're also thinking this would be a more comprehensive approach, I'll do some checking and then make the change.

Reply:
A good fix, even it's a best effort.

Is it possible to get the IM resource and set the status of stopped instances to stoped during the IM controller initialization?

Reply:
> Is it possible to get the IM resource and set the status of stopped instances to stopped during the IM controller initialization?

I am thinking it would be better to do it in the normal reconciliation loop, since (I believe) it is possible for a second longhorn-manager's instance manager controller to take responsibility for one node's instance manager in some cases, and we would still like this state to be zeroed out appropriately.

I was thinking about adding something in handlePod to ensure we never start an instance manager without zeroing out the status, but the combination of [this check](https://github.com/longhorn/longhorn-manager/blob/8457624bfffd73f4a56610b90c1b3e81c26c4554/controller/instance_manager_controller.go#L438-L440) and @PhanLe1010's `InstanceManagerStateError` suggestion make this unnecessary. (We will never start an instance manager unless we are in `InstanceManagerStateStopped` or `InstanceManagerStateError`, and implementing his suggestion guarantees we will have already zeroed out the status.)

Reply:
Well explained. Sounds good to me.